File: src/models/Order.ts
import mongoose from 'mongoose';

const orderSchema = new mongoose.Schema({
    products: [
        {
            productId: {
                type: mongoose.Schema.Types.ObjectId,
                ref: 'Product',
                required: true
            },
            name: String,
            price: Number,
            quantity: {
                type: Number,
                default: 1
            }
        }
    ],
    orderNumber: {
        type: Number,
        required: true,
        unique: true
    },
    totalAmount: {
        type: Number,
        required: true
    },
    customerName: {
        type: String,
        required: false // Optional for initial draft
    },
    customerPhone: {
        type: String,
        required: false // Optional for very early drafts
    },
    customerAddress: {
        type: String,
        required: false // Optional for draft orders
    },
    couponCode: {
        type: String,
        default: null
    },
    discountAmount: {
        type: Number,
        default: 0
    },
    paymentMethod: {
        type: String,
        enum: ['nagad', 'rocket', 'cod'],
        default: 'cod'
    },
    shippingZone: {
        type: String,
        default: 'dhaka'
    },
    shippingCost: {
        type: Number,
        default: 60
    },
    consignmentId: {
        type: String,
        default: null
    },
    pathaoStatus: {
        type: String,
        default: null
    },
    status: {
        type: String,
        enum: ['incomplete', 'pending', 'confirmed', 'shipped', 'delivered', 'cancelled'],
        default: 'pending'
    },
    createdAt: {
        type: Date,
        default: Date.now
    }
});
// @database-optimizer: Add indexes for common query patterns
orderSchema.index({ orderNumber: 1 }, { unique: true }); // Fast order lookup by number
orderSchema.index({ customerPhone: 1 }); // Customer order history by phone
orderSchema.index({ status: 1, createdAt: -1 }); // Admin filter by status + sort
orderSchema.index({ createdAt: -1 }); // Sort by newest (admin dashboard)
orderSchema.index({ consignmentId: 1 }); // Pathao tracking lookup
// Index for cleaning up old incomplete orders easily
orderSchema.index({ createdAt: 1 }, { expireAfterSeconds: 14 * 24 * 60 * 60, partialFilterExpression: { status: 'incomplete' } });

const Order = mongoose.models.Order || mongoose.model('Order', orderSchema);

export default Order;

File: ecom_app/lib/models/order.dart
class Order {
  final String? id;
  final List<OrderItem> products;
  final int? orderNumber;
  final double totalAmount;
  final String customerName;
  final String customerPhone;
  final String customerAddress;
  final String? couponCode;
  final double discountAmount;
  final String paymentMethod;
  final String shippingZone;
  final double shippingCost;
  final String status;
  final String? consignmentId;
  final String? pathaoStatus;
  final DateTime? createdAt;

  Order({
    this.id,
    required this.products,
    this.orderNumber,
    required this.totalAmount,
    required this.customerName,
    required this.customerPhone,
    required this.customerAddress,
    this.couponCode,
    this.discountAmount = 0,
    this.paymentMethod = 'cod',
    this.shippingZone = 'dhaka',
    this.shippingCost = 60,
    this.status = 'pending',
    this.consignmentId,
    this.pathaoStatus,
    this.createdAt,
  });

  factory Order.fromJson(Map<String, dynamic> json) {
    return Order(
      id: json['_id'],
      orderNumber: json['orderNumber'],
      products:
          (json['products'] as List?)
              ?.map((p) => OrderItem.fromJson(p))
              .toList() ??
          [],
      totalAmount: (json['totalAmount'] ?? 0).toDouble(),
      customerName: json['customerName'] ?? '',
      customerPhone: json['customerPhone'] ?? '',
      customerAddress: json['customerAddress'] ?? '',
      couponCode: json['couponCode'],
      discountAmount: (json['discountAmount'] ?? 0).toDouble(),
      paymentMethod: json['paymentMethod'] ?? 'cod',
      shippingZone: json['shippingZone'] ?? 'dhaka',
      shippingCost: (json['shippingCost'] ?? 0).toDouble(),
      status: json['status'] ?? 'pending',
      consignmentId: json['consignmentId'],
      pathaoStatus: json['pathaoStatus'],
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'products': products.map((x) => x.toJson()).toList(),
      'customerName': customerName,
      'customerPhone': customerPhone,
      'customerAddress': customerAddress,
      'couponCode': couponCode,
      'discountAmount': discountAmount,
      'paymentMethod': paymentMethod,
      'shippingZone': shippingZone,
      'shippingCost': shippingCost,
      'consignmentId': consignmentId,
      'pathaoStatus': pathaoStatus,
    };
  }
}

class OrderItem {
  final String productId;
  final String? name;
  final double? price;
  final int quantity;

  OrderItem({
    required this.productId,
    this.name,
    this.price,
    required this.quantity,
  });

  factory OrderItem.fromJson(Map<String, dynamic> json) {
    return OrderItem(
      productId: json['productId'] ?? '',
      name: json['name'],
      price: (json['price'] ?? 0).toDouble(),
      quantity: json['quantity'] ?? 1,
    );
  }

  Map<String, dynamic> toJson() {
    return {'productId': productId, 'quantity': quantity};
  }
}
